### DISCUSSION TITLE
Fail to find matching pair of PMTF/A23

### DISCUSSION THREAD

**Discussion by Ho-ChunHuang-NOAA** [dtcenter/METplus#1833] [Statistical Computation] [answered]
I have a grib2 file with 3 days of forecast.  The point_stat failed because of not finding the PMTF/A23 from this forecasting output.  The $WGRIB2 command and degrib2 command seem to indicate the time period is correct so I need help of why it can not find the record with description of "PMTF/A23"  Thank you.

Error log output
[master_metplus.log.20220923190853.log](https://github.com/dtcenter/METplus/files/9636839/master_metplus.log.20220923190853.log)

Can not upload grib2 file so please go to
https://www.emc.ncep.noaa.gov/mmb/hchuang/ftp/ to get
aqm.t06z.ave_24hr_pm25.793.grib2

It is the /lfs/h2/emc/ptmp/ho-chun.huang/com/aqm/v70c3/cs.20220815/aqm.t06z.ave_24hr_pm25.793.grib2 in the error log file.

**Comment by JohnHalleyGotway** [comment]
@Ho-ChunHuang-NOAA, I see you have a question about configuring METplus to verify PMTF data. Thanks for posting your sample data to the FTP site. That makes it much easier to debug. I grabbed that data and see what's going on here.

Here's the output from wgrib2:
```
wgrib2 aqm.t06z.ave_24hr_pm25.793.grib2 
1:0:d=2022081506:PMTF:1 hybrid level:-2147483647-22 hour ave fcst:
2:172095:d=2022081506:PMTF:1 hybrid level:23-46 hour ave fcst:
3:376183:d=2022081506:PMTF:1 hybrid level:47-70 hour ave fcst:
```
Note that the second record contains an **AVERAGE** computed between hours **23 and 46**. And the third record is between hours **47 and 70**. Both records 2 and 3 are computed over a 23 hour averaging period. Because of the way this data is stored in the GRIB2 file, requesting "A23" will match both of these records, but they're just valid at different times. You can see that by running plot_data_plane:
```
plot_data_plane aqm.t06z.ave_24hr_pm25.793.grib2 plot.ps 'name="PMTF";level="A23";'
```
That produces this warning message:
```
WARNING: MetGrib2DataFile::data_plane() -> Using the first of 2 matching records for "PMTF/A23":
WARNING: record 2 field 1: ipdtmpl[29] = 13, 193, 2, 0, 89, 0, 0, 1, 23, 105, 0, 1, 255, -127, -2147483647, 2022, 8, 17, 4, 0, 0, 1, 0, 0, 2, 1, 23, 255, 0
WARNING: record 3 field 1: ipdtmpl[29] = 13, 193, 2, 0, 89, 0, 0, 1, 47, 105, 0, 1, 255, -127, -2147483647, 2022, 8, 18, 4, 0, 0, 1, 0, 0, 2, 1, 23, 255, 0
```
I see in your log file that you're specifying a valid time, like this:
```
FCST_FIELD={ name="PMTF"; level="A23"; valid_time= "20220816_04"; set_attr_name = "PMAVE"; }
```
That *should* work, but the valid time you're requesting doesn't actually match the data in the file. See the valid time of the data by running wgrib2 with the `-vt` command line option (-vt prints the verification time):
```
wgrib2 -vt  aqm.t06z.ave_24hr_pm25.793.grib2 
1:0:vt=2022081604
2:172095:vt=2022081704
3:376183:vt=2022081804
```
You requested the 16th, but the 2nd and 3rd records are for the 17th and 18th. Note that specifying the `lead_time` is another alternative that would also work. So the following 2 plot_data_plane commands both match the 2nd record in that file:
```
plot_data_plane aqm.t06z.ave_24hr_pm25.793.grib2 plot.ps 'name="PMTF";level="A23";valid_time="20220817_04";'
plot_data_plane aqm.t06z.ave_24hr_pm25.793.grib2 plot.ps 'name="PMTF";level="A23";lead_time="46";'
```
If you're trying to match the first record, that's difficult. This data is not stored well in the GRIB2 file:
```
1:0:d=2022081506:PMTF:1 hybrid level:-2147483647-22 hour ave fcst:
```
It's specified as an accumulation, but the first time of that accumulation is NOT specified and has a value of -2147483647. Is that the one you're actually trying to read?

**Comment by Ho-ChunHuang-NOAA** [comment, marked as answer]
Hi, John:

Thanks.

Yes, the first record is FCST hours  -1 to 22, so we have 04z to 04z for
the entire day.  I tried to run the verification for the date of 20220916
it will find the first record, verification date of, 20220917 it will find
the 2nd record, but verification date of 20220918, it can not find the
third record.  I don't really know what happened.

If I skip A23 and simply use L1, will the 3 record been found one by one?

Ho-Chun Huang, Ph.D.

**Reply by JohnHalleyGotway** [reply]
@Ho-ChunHuang-NOAA I did some more digging and would recommend using the following settings to match each of these records:
1. FCST_FIELD={ name="PMTF"; level="L1"; valid_time= "20220816_04"; set_attr_name = "PMAVE"; }
2. FCST_FIELD={ name="PMTF"; level="L1"; valid_time= "20220817_04"; set_attr_name = "PMAVE"; }
3. FCST_FIELD={ name="PMTF"; level="L1"; valid_time= "20220818_04"; set_attr_name = "PMAVE"; }

Please let me know if your run into additional issues.

**Reply by Ho-ChunHuang-NOAA** [reply]
Hi, John:

With the criteria suggested by you, the code/script can read the first record of the grib2 file.  I will go back to the author of the grib2 files and see if they can fix the grib2 encoding of the 1st record.  Otherwise, we can use L1 as an alternative choice.

Please close this discussion.

Thank you very much for your help.
